src/lib.rs

```diff
@@ -46,49 +52,108 @@ impl Clipboard {
         Self { request_receiver, request_sender, clipboard_thread }
     }
 
+    /// Load custom clipboard data.
+    ///
+    /// Load the requested type from a clipboard on the last observed seat.
+    pub fn load<T: AllowedMimeTypes + 'static>(&self) -> Result<T> {
+        self.load_inner(SelectionTarget::Clipboard, T::allowed())
+    }
+
     /// Load clipboard data.
     ///
     /// Loads content from a clipboard on a last observed seat.
-    pub fn load(&self) -> Result<String> {
-        let _ = self.request_sender.send(worker::Command::Load);
-
-        if let Ok(reply) = self.request_receiver.recv() {
-            reply
-        } else {
-            // The clipboard thread is dead, however we shouldn't crash downstream, so
-            // propogating an error.
-            Err(std::io::Error::new(std::io::ErrorKind::Other, "clipboard is dead."))
-        }
+    pub fn load_text(&self) -> Result<String> {
+        self.load::<Text>().map(|t| t.0)
     }
 
-    /// Store to a clipboard.
+    /// Load custom primary clipboard data.
     ///
-    /// Stores to a clipboard on a last observed seat.
-    pub fn store<T: Into<String>>(&self, text: T) {
-        let request = worker::Command::Store(text.into());
-        let _ = self.request_sender.send(request);
+    /// Load the requested type from a primary clipboard on the last observed
+    /// seat.
+    pub fn load_primary<T: AllowedMimeTypes + 'static>(&self) -> Result<T> {
+        self.load_inner(SelectionTarget::Primary, T::allowed())
     }
 
     /// Load primary clipboard data.
     ///
     /// Loads content from a  primary clipboard on a last observed seat.
-    pub fn load_primary(&self) -> Result<String> {
-        let _ = self.request_sender.send(worker::Command::LoadPrimary);
-
-        if let Ok(reply) = self.request_receiver.recv() {
-            reply
-        } else {
-            // The clipboard thread is dead, however we shouldn't crash downstream, so
-            // propogating an error.
-            Err(std::io::Error::new(std::io::ErrorKind::Other, "clipboard is dead."))
-        }
+    pub fn load_primary_text(&self) -> Result<String> {
+        self.load_primary::<Text>().map(|t| t.0)
+    }
+
+    /// Load raw clipboard data.
+    ///
+    /// Loads content from a  primary clipboard on a last observed seat.
+    pub fn load_raw(
+        &self,
+        allowed: impl Into<Cow<'static, [MimeType]>>,
+    ) -> Result<(Vec<u8>, MimeType)> {
+        self.load_inner(SelectionTarget::Clipboard, allowed)
+    }
+
+    /// Load raw primary clipboard data.
+    ///
+    /// Loads content from a  primary clipboard on a last observed seat.
+    pub fn load_primary_raw(
+        &self,
+        allowed: impl Into<Cow<'static, [MimeType]>>,
+    ) -> Result<(Vec<u8>, MimeType)> {
+        self.load_inner(SelectionTarget::Primary, allowed)
+    }
```

Comment:
What's the use case for `raw`?

Reply:
I found it useful to have an option without trait bounds on the return type, when passing around a boxed trait object clipboard. I still use the `AllowedMimeTypes` trait, but not in the return type trait for the clipboard because Boxing it in a return type doesn't seem to be object safe.

Reply:
If the trait were allowed to take an optional reference to self like below, the `raw` methods could be avoided, but the api would probably need to include an argument Option<&(impl AllowedMimeTypes + 'static)>. This way users could create their own type type implementing AllowedMimeTypes which just wraps the received raw data and the allowed mime types.

```rust

/// Describes the mime types which are accepted.
pub trait AllowedMimeTypes: TryFrom<(Vec<u8>, MimeType)> {
    /// List allowed mime types for the type to convert from a byte slice.
    ///
    /// Allowed mime types should be listed in order of decreasing preference,
    /// most preferred first.
    fn allowed(_self: Option<&self>) -> Cow<'static, [MimeType]>;
}
```